packages/node/test/retry.test.ts

```diff
@@ -0,0 +1,107 @@
+import { TestRetry, MOCK_MAX_RETRIES } from './mocks/retry';
+import * as nock from 'nock';
+import { Event, Status } from '@amplitude/types';
+import { asyncSleep } from '@amplitude/utils';
+
+import { AMPLITUDE_SERVER_URL } from '../src/constants';
+
+const FAILING_USER_ID = 'data_monster';
+const PASSING_USER_ID = 'node_monster';
+
+const generateEvent = (userId: string): Event => {
+  return {
+    event_id: 0,
+    user_id: userId,
+    event_type: 'NOT_A_REAL_EVENT',
+  };
+};
+
+describe('retry mechanisms layer', () => {
+  // A helper that persistently listens to nock and returns the # of
+  // times a user id has been included and hasn't been included.
+  let nockHook = { matchPassCount: 0, matchFailCount: 0 };
+  beforeEach(() => {
+    nock(AMPLITUDE_SERVER_URL)
+      .persist()
+      .post(
+        () => true,
+        body => {
+          const events: Array<Event> = body.events ?? [];
+          const isMatch = events.some(event => event.user_id === FAILING_USER_ID);
+
+          if (isMatch) {
+            nockHook.matchPassCount += 1;
+            return true;
+          } else {
+            nockHook.matchFailCount += 1;
+            return false;
+          }
+        },
+      )
+      .reply(400);
+
+    nock(AMPLITUDE_SERVER_URL)
+      .persist()
+      .post(
+        () => true,
+        body => {
+          const events: Array<Event> = body.events ?? [];
+          return !events.some(event => event.user_id === FAILING_USER_ID);
+        },
+      )
+      .reply(200);
+  });
+  // After each test, remove any remaining mocks and reset the nock hook.
+  afterEach(() => {
+    nock.cleanAll();
+    nockHook = {
+      matchPassCount: 0,
+      matchFailCount: 0,
+    };
+  });
+
+  it('should not retry events that pass', async () => {
+    const payload = [generateEvent(PASSING_USER_ID)];
+
+    const retryHandler = new TestRetry();
+
+    const response = await retryHandler.sendEventsWithRetry(payload);
+
+    expect(response.status).toBe(Status.Success);
+    expect(response.statusCode).toBe(200);
+    // One response goes out matching the initial send
+    expect(nockHook.matchFailCount).toBe(1);
+  });
+
+  it('should retry events that fail', async () => {
+    const payload = [generateEvent(FAILING_USER_ID)];
+
+    const retryHandler = new TestRetry();
+
+    const response = await retryHandler.sendEventsWithRetry(payload);
+
+    await asyncSleep(1000);
+
+    expect(response.status).toBe(Status.Invalid);
+    expect(response.statusCode).toBe(400);
```

Comment:
Ah, there's still the issue of retrying bad requests. We should not be retrying requests for events that trigger 400 errors.